File: src/infrastructure/memory/memoryRepositories.ts
import type {
  Id,
  ImportBatch,
  SourceRecord,
  Transaction,
} from '../../domain/entities/types'
import type {
  ImportBatchRepository,
  SourceRecordRepository,
  TransactionRepository,
} from '../../application/ports/repositories'

/**
 * مستودعات في الذاكرة — للاختبار من fixtures (CLAUDE.md #6 و #15).
 *
 * تُستخدم لتشغيل كل حالات spec/06 ومعايير OVERRIDES §7
 * **بدون فايربيز إطلاقًا**، وهو ما يجعل معيار القبول قابلًا للتحقق أصلًا.
 *
 * النسخ عند القراءة والكتابة مقصود: يمنع تعديل الحالة من خارج المستودع،
 * فيتصرف كقاعدة بيانات حقيقية لا كمصفوفة مشتركة.
 */

const clone = <T>(value: T): T => structuredClone(value)

/**
 * مستودع يقدر يرجّع حالته لِما كانت عليه.
 * هو ما يجعل MemoryUnitOfWork ذرّية فعليًا لا اسمًا،
 * ويثبت حالة spec/06: «انقطاع أثناء حفظ دفعة ⇒ صفر أو كامل الدفعة».
 */
export interface Snapshotable<S = unknown> {
  snapshot(): S
  restore(state: S): void
}

/** أساس مشترك لمستودعات الذاكرة المفتاحية. */
abstract class KeyedStore<T> implements Snapshotable<Map<Id, T>> {
  protected items = new Map<Id, T>()

  snapshot(): Map<Id, T> {
    // نسخة عميقة: التراجع لازم يرجّع محتوى المستندات لا مفاتيحها فقط
    return new Map([...this.items].map(([k, v]) => [k, clone(v)]))
  }

  restore(state: Map<Id, T>): void {
    this.items = new Map([...state].map(([k, v]) => [k, clone(v)]))
  }
}

export class MemoryTransactionRepository extends KeyedStore<Transaction> implements TransactionRepository {

  async listByDateRange(fromIso: string, toIso: string): Promise<Transaction[]> {
    return [...this.items.values()]
      .filter((t) => t.occurredAt >= fromIso && t.occurredAt <= toIso)
      .sort((a, b) =>
        a.occurredAt === b.occurredAt
          ? a.sourceOrder - b.sourceOrder
          : a.occurredAt < b.occurredAt
            ? -1
            : 1,
      )
      .map(clone)
  }

  async listByBatch(): Promise<Transaction[]> {
    return [] // العلاقة بالدفعة عبر SourceRecord — ليست ملكية مطلقة (spec/03)
  }

  async findByIds(ids: readonly Id[]): Promise<Transaction[]> {
    return ids.map((id) => this.items.get(id)).filter((t): t is Transaction => !!t).map(clone)
  }

  async saveMany(transactions: readonly Transaction[]): Promise<void> {
    for (const t of transactions) this.items.set(t.id, clone(t))
  }

  async update(id: Id, patch: Partial<Transaction>): Promise<void> {
    const existing = this.items.get(id)
    if (!existing) throw new Error(`عملية غير موجودة: ${id}`)
    this.items.set(id, { ...existing, ...clone(patch), id })
  }

  async deleteMany(ids: readonly Id[]): Promise<void> {
    for (const id of ids) this.items.delete(id)
  }

  /** للاختبار فقط — عدد ما هو مخزَّن فعلًا. */
  size(): number {
    return this.items.size
  }

  all(): Transaction[] {
    return [...this.items.values()].map(clone)
  }
}

export class MemorySourceRecordRepository extends KeyedStore<SourceRecord> implements SourceRecordRepository {

  async listByBatch(batchId: Id): Promise<SourceRecord[]> {
    return [...this.items.values()].filter((r) => r.batchId === batchId).map(clone)
  }

  async listByAccountIdentity(accountIdentity: string): Promise<SourceRecord[]> {
    return [...this.items.values()]
      .filter((r) => r.accountIdentity === accountIdentity)
      .map(clone)
  }

  async listByTransactionIds(ids: readonly Id[]): Promise<SourceRecord[]> {
    const wanted = new Set(ids)
    return [...this.items.values()]
      .filter((r) => r.transactionId !== null && wanted.has(r.transactionId))
      .map(clone)
  }

  async saveMany(records: readonly SourceRecord[]): Promise<void> {
    for (const r of records) this.items.set(r.id, clone(r))
  }

  async deleteMany(ids: readonly Id[]): Promise<void> {
    for (const id of ids) this.items.delete(id)
  }

  all(): SourceRecord[] {
    return [...this.items.values()].map(clone)
  }
}

export class MemoryImportBatchRepository extends KeyedStore<ImportBatch> implements ImportBatchRepository {

  async findById(id: Id): Promise<ImportBatch | null> {
    const found = this.items.get(id)
    return found ? clone(found) : null
  }

  async findByFileHash(fileHash: string): Promise<ImportBatch | null> {
    const found = [...this.items.values()].find(
      (b) => b.fileHash === fileHash && b.state === 'committed',
    )
    return found ? clone(found) : null
  }

  async listRecent(limit: number): Promise<ImportBatch[]> {
    return [...this.items.values()]
      .sort((a, b) => (a.importedAt < b.importedAt ? 1 : -1))
      .slice(0, limit)
      .map(clone)
  }

  async save(batch: ImportBatch): Promise<void> {
    this.items.set(batch.id, clone(batch))
  }

  async updateState(id: Id, state: ImportBatch['state']): Promise<void> {
    const existing = this.items.get(id)
    if (!existing) throw new Error(`دفعة غير موجودة: ${id}`)
    this.items.set(id, { ...existing, state })
  }

  all(): ImportBatch[] {
    return [...this.items.values()].map(clone)
  }
}


export * from './memoryReferenceRepositories'
export * from './memorySupport'

File: native-app/app/src/commonMain/kotlin/app/masroufy/memory/MemoryRepositories.kt
package app.masroufy.memory

import app.masroufy.core.Budget
import app.masroufy.core.Category
import app.masroufy.core.CategoryBudget
import app.masroufy.core.Id
import app.masroufy.core.PersonAllocation
import app.masroufy.core.Transaction
import app.masroufy.port.AllocationRepository
import app.masroufy.port.BudgetRepository
import app.masroufy.port.CategoryRepository
import app.masroufy.port.TransactionPatch
import app.masroufy.port.TransactionRepository

/**
 * مستودعات في الذاكرة — التنفيذ التاني لكل واجهة (CLAUDE.md #6)، للاختبار من fixtures.
 * بتقلّد تنفيذ التطبيق الحالي (`src/infrastructure/memory/`) **بنفس الترتيب والفلترة بالظبط**،
 * عشان نتيجة حالة الاستخدام تطابق ملف المرجع بالحرف.
 *
 * الكيانات في كوتلن `data class` (مش بتتغير)، فمفيش نسخ زي `clone` بتاعة جافاسكربت.
 */
class MemoryTransactionRepository(seed: List<Transaction> = emptyList()) : TransactionRepository, Snapshotable {
    private var items = LinkedHashMap<Id, Transaction>()

    init {
        for (t in seed) items[t.id] = t
    }

    /** نفس ترتيب التطبيق الحالي: التاريخ تصاعدي، وبعده `sourceOrder`. */
    override suspend fun listByDateRange(fromIso: String, toIso: String): List<Transaction> =
        items.values.filter { it.occurredAt >= fromIso && it.occurredAt <= toIso }
            .sortedWith(compareBy({ it.occurredAt }, { it.sourceOrder }))

    override suspend fun listCreatedAfter(iso: String): List<Transaction> =
        items.values.filter { it.createdAt > iso }.sortedBy { it.createdAt }

    /** العلاقة بالدفعة بتيجي من `SourceRecord` مش من العملية نفسها — زي التطبيق الحالي (spec/03). */
    override suspend fun listByBatch(batchId: Id): List<Transaction> = emptyList()

    override suspend fun findByIds(ids: List<Id>): List<Transaction> = ids.mapNotNull { items[it] }

    override suspend fun saveMany(transactions: List<Transaction>) {
        for (t in transactions) items[t.id] = t
    }

    override suspend fun update(id: Id, patch: TransactionPatch) {
        val old = items[id] ?: return
        items[id] = old.copy(
            amountMinor = patch.amountMinor ?: old.amountMinor,
            originalAmountMinor = patch.originalAmountMinor ?: old.originalAmountMinor,
            economicKind = patch.economicKind ?: old.economicKind,
            economicKindConfirmed = patch.economicKindConfirmed ?: old.economicKindConfirmed,
            categoryId = if (patch.clearCategoryId) null else patch.categoryId ?: old.categoryId,
            categoryConfirmed = patch.categoryConfirmed ?: old.categoryConfirmed,
            reviewState = patch.reviewState ?: old.reviewState,
            merchantId = patch.merchantId ?: old.merchantId,
            note = if (patch.clearNote) null else patch.note ?: old.note,
            walletId = patch.walletId ?: old.walletId,
            excludedFromBudget = patch.excludedFromBudget ?: old.excludedFromBudget,
            isCashTagged = patch.isCashTagged ?: old.isCashTagged,
            updatedAt = patch.updatedAt ?: old.updatedAt,
        )
    }

    override suspend fun deleteMany(ids: List<Id>) {
        for (id in ids) items.remove(id)
    }

    fun all(): List<Transaction> = items.values.toList()

    override fun snapshot(): Any = LinkedHashMap(items)

    @Suppress("UNCHECKED_CAST")
    override fun restore(state: Any) {
        items = LinkedHashMap(state as LinkedHashMap<Id, Transaction>)
    }
}

class MemoryCategoryRepository(seed: List<Category> = emptyList()) : CategoryRepository {
    private val items = LinkedHashMap<Id, Category>()

    init {
        for (c in seed) items[c.id] = c
    }

    override suspend fun listAll(): List<Category> = items.values.toList()

    override suspend fun save(category: Category) {
        items[category.id] = category
    }
}

class MemoryAllocationRepository(seed: List<PersonAllocation> = emptyList()) : AllocationRepository {
    private val items = LinkedHashMap<Id, PersonAllocation>()

    init {
        for (a in seed) items[a.id] = a
    }

    override suspend fun listByTransactionIds(ids: List<Id>): List<PersonAllocation> {
        val wanted = ids.toSet()
        return items.values.filter { it.transactionId in wanted }
    }

    override suspend fun saveMany(allocations: List<PersonAllocation>) {
        for (a in allocations) items[a.id] = a
    }

    override suspend fun deleteMany(ids: List<Id>) {
        for (id in ids) items.remove(id)
    }
}

class MemoryBudgetRepository(seed: List<Budget> = emptyList(), lines: List<CategoryBudget> = emptyList()) : BudgetRepository {
    private val budgets = LinkedHashMap<String, Budget>()
    private val categoryLines = LinkedHashMap<Id, CategoryBudget>()

    init {
        for (b in seed) budgets[b.periodKey] = b
        for (l in lines) categoryLines[l.id] = l
    }

    override suspend fun findByPeriod(periodKey: String): Budget? = budgets[periodKey]

    override suspend fun save(budget: Budget) {
        budgets[budget.periodKey] = budget
    }

    override suspend fun remove(periodKey: String) {
        val removed = budgets.remove(periodKey) ?: return
        categoryLines.values.filter { it.budgetId == removed.id }.forEach { categoryLines.remove(it.id) }
    }

    override suspend fun listCategoryBudgets(budgetId: Id): List<CategoryBudget> =
        categoryLines.values.filter { it.budgetId == budgetId }

    override suspend fun saveCategoryBudget(line: CategoryBudget) {
        categoryLines[line.id] = line
    }

    override suspend fun removeCategoryBudget(id: Id) {
        categoryLines.remove(id)
    }
}
